Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,748 @@
+            }
+            if task.countOfBytesExpectedToReceive == task.countOfBytesReceived {
+                fileHandle.closeFile()
+                self.properties[.temporaryFileURL] = self.tempFileURL
+            }
+        }
+    }
+
+    fileprivate func notifyDelegate(aboutUploadedData count: Int64) {
+        guard let task = self.task as? URLSessionUploadTask,
+            let session = self.task?.session as? URLSession,
+            case .taskDelegate(let delegate) = session.behaviour(for: task) else { return }
+        task.countOfBytesSent += count
+        session.delegateQueue.addOperation {
+            delegate.urlSession(session, task: task, didSendBodyData: count,
+                                totalBytesSent: task.countOfBytesSent, totalBytesExpectedToSend: task.countOfBytesExpectedToSend)
+        }
+    }
+    
+    func didReceive(headerData data: Data, contentLength: Int64) -> _EasyHandle._Action {
+        NSRequiresConcreteImplementation()
+    }
+
+    func fill(writeBuffer buffer: UnsafeMutableBufferPointer<Int8>) -> _EasyHandle._WriteBufferResult {
+        guard case .transferInProgress(let ts) = internalState else {
+            fatalError("Requested to fill write buffer, but transfer isn't in progress.")
+        }
+        guard let source = ts.requestBodySource else {
+            fatalError("Requested to fill write buffer, but transfer state has no body source.")
+        }
+        switch source.getNextChunk(withLength: buffer.count) {
+        case .data(let data):
+            copyDispatchData(data, infoBuffer: buffer)
+            let count = data.count
+            assert(count > 0)
+            notifyDelegate(aboutUploadedData: Int64(count))
+            return .bytes(count)
+        case .done:
+            return .bytes(0)
+        case .retryLater:
+            // At this point we'll try to pause the easy handle. The body source
+            // is responsible for un-pausing the handle once data becomes
+            // available.
+            return .pause
+        case .error:
+            return .abort
+        }
+    }
+
+    func transferCompleted(withErrorCode errorCode: Int?) {
+        // At this point the transfer is complete and we can decide what to do.
+        // If everything went well, we will simply forward the resulting data
+        // to the delegate. But in case of redirects etc. we might send another
+        // request.
+        guard case .transferInProgress(let ts) = internalState else {
+            fatalError("Transfer completed, but it wasn't in progress.")
+        }
+        guard let request = task?.currentRequest else {
+            fatalError("Transfer completed, but there's no current request.")
+        }
+        guard errorCode == nil else {
+            internalState = .transferFailed
+            failWith(errorCode: errorCode!, request: request)
+            return
+        }
+
+        if let response = task?.response {
```

Comment:
Why not just `ts.response = task?.response` (and change `let ts` to `var ts` above).